walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,71 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // timeIsOccupied is true if someone in the request attendees has an event in that moment
+    boolean[] timeIsOccupied = new boolean[TimeRange.END_OF_DAY + 1];
+    Collection<String> requestAttendees = request.getAttendees();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      for (String s : requestAttendees) {
+        // if someone in our request is in this event,
+        // make the range of the event [start,end) true in the timeIsOccupied[]
+        if (eventAttendees.contains(s)) {
+          TimeRange eventRange = event.getWhen();
+          for (int i = eventRange.start(); i < eventRange.end(); ++i) {
+            timeIsOccupied[i] = true;
+          }
+          break;
+        }
+      }
+    }
+    return findResultOfQuery(timeIsOccupied, request.getDuration());
+  }
+  /**
+   * finds continous intervals of false in timeIsOcuppied[] with lengths greater than or equal to
+   * durationOfRequest and returns them as a Collection<TimeRange>
+   */
+  private Collection<TimeRange> findResultOfQuery(
+      boolean[] timeIsOccupied, long durationOfRequest) {
+    Collection<TimeRange> result = new ArrayList<TimeRange>();
+    int start = 0;
+    int end = 1;
+    while (end < timeIsOccupied.length) {
+      // if start is occupied move forward to search for a valid start
+      if (timeIsOccupied[start]) {
+        start++;
+        end = start + 1;
+        // start is not occupied (valid) but end is occupied
+        // meaning that there is range of falses [start,end)
+      } else if (timeIsOccupied[end]) {
+        TimeRange validRange = TimeRange.fromStartEnd(start, end, false);
```

Comment:
It's not clear what this `false` is here. To make it clearer, it's common to add a comment with the name of the parameter: `TimeRange validRange = TimeRange.fromStartEnd(start, end, /*inclusive=*/ false);`